rsconnect/validation.py

```diff
@@ -1,46 +1,121 @@
 import typing
 
+import click
+
 from rsconnect.exception import RSConnectException
 
 
-def _get_present_options(options: typing.Dict[str, typing.Optional[str]]) -> typing.List[str]:
-    return [k for k, v in options.items() if v]
+def _get_present_options(
+    options: typing.Dict[str, typing.Optional[typing.Any]],
+    ctx: click.Context,
+) -> typing.List[str]:
+    result: typing.List[str] = []
+    for k, v in options.items():
+        if v:
+            parts = k.split("--")
+            if ctx and len(parts) == 2:
+                varName = parts[1].replace("-", "_")
+                source = ctx.get_parameter_source(varName)  # type: ignore
```

Comment:
This function was introduced in Click 8.0. Now that our minimum python version is 3.7, we can require click >= 8.0.